### ISSUE TITLE
[Feature request]: Accepting gen AI code seems to be potentially a bad idea

### ISSUE THREAD

**Issue by ell1e** [flatpak/flatpak#6509]
### Checklist

- [x] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

Accepting gen AI code seems to be potentially a bad idea due to the [plagiarism problem](https://github.com/Vxrpenter/AIMania/blob/main/WHY.md#extensive-plagiarism), also see here:

https://lcamtuf.substack.com/p/large-language-models-and-plagiarism

> Bard didn’t merely copy facts when composing its answer; it lifted a good chunk of the text wholesale — wording, parentheses, non-US units, and all.

https://dl.acm.org/doi/10.1145/3543507.3583199

> Our results suggest that [...] three types of plagiarism widely exist in LMs beyond memorization, [...] Given that a majority of LMs’ training data is scraped from the Web without informing content owners, their reiteration of words, phrases, and even core ideas from training sets into generated texts has ethical implications. Their patterns are likely to exacerbate as both the size of LMs and their training data increase, [...] Plagiarized content can also contain individuals’ personal and sensitive information.

https://archive.is/1EzVK

> I reached out to the team about this. Apparently all public GitHub code was used in training. We don't distinguish by license type.

The available info on this topic suggests that for all major code models the training data typically doesn't seem to be 100% licensed in some way that doesn't require attribution, and at the same time plagiarism in the output of the training data seems to be the rule rather than to be the exception.

Therefore, it seems like the natural conclusion that gen AI code commits may typically violate the GPL and other licenses. Since I recently saw somebody use Claude for flatpak commits, I thought I would bring it up.

PS: I'm not a lawyer, this isn't legal advice. Check the sources for yourself to make up your mind.

**Comment by ghost** [comment]
It will bring more low quality PR's and issues which has caused issues for open source projects like [curl](https://www.theregister.com/2026/01/21/curl_ends_bug_bounty/). Also AI costs a lot of [energy and water](https://www.cbsnews.com/chicago/news/data-centers-for-ai-electricity-water-climate-health/). And research points out usage of AI leads to worse [productivity](https://arxiv.org/abs/2510.10165) and [decreased critical thinking](https://ctl.duke.edu/ai-ethics-learning-toolkit/does-ai-harm-critical-thinking/). Also [ChatGPT](https://github.com/bobmatyas/wp-block-ai-crawlers/issues/110#issuecomment-3924068295) (and [others](https://appleinsider.com/articles/25/08/05/perplexity-defensive-over-ignoring-robotstxt-and-stealing-data)) have and still ignore the robots.txt to prevent scraping.

**Comment by linghengqian** [comment]
This sounds more like getting flatpak to refuse to use code generated by github copilot. Simply making a PR look like it wasn't written by AI isn't difficult; there are many local AI agents that can further obfuscate PR content. 

From my perspective, it really depends more on the preferences of the project maintainers. In the ASF next door, a large number of projects have `CLAUDE.md`, `AGENTS.md`, and `.codex/skills`. See https://github.com/search?q=org%3Aapache%20%22CLAUDE.md%22&type=code . It's simply a matter of contributors using AI to generate code and self-reviewing it, while maintainers use AI to review pull requests.

I'm not a lawyer, but if this is the final opinion, I think it can be written into https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md .

**Comment by ell1e** [comment]
>  refuse to use code generated by github copilot.

Claude:

https://privacy.claude.com/en/articles/10023555-how-do-you-use-personal-data-in-model-training

> The following sources of training data may contain personal data: [...] Publicly available information via the Internet

Pick any of the big code model you like, and I can probably find a confirmation that they use GPL code.

**Comment by linghengqian** [comment]
Considering the steps outlined in https://github.com/rasbt/LLMs-from-scratch , I have absolutely no belief that the model training process would intentionally prevent the model from reading GPL code. Models mimicking humans to avoid being identified as AI is a rapidly expanding trend, occurring even without GitHub, and in places like GitLab or Gitea. Personally, I think such constraints will remain unrealistic; sometimes I'm not even sure if the person opening the PR is a human or something like OpenClaw. 

That said, you can still directly open and update the PR at https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md .

**Comment by ell1e** [comment]
> I have absolutely no belief that the model training process would intentionally prevent the model from reading GPL code

I agree. That is why I brought up the issue.

**Comment by ell1e** [comment]
> sometimes I'm not even sure if the person opening the PR is a human or something like OpenClaw.

Even then, I feel like having clear rules could help making people stop that (and I'm not a lawyer, but I wonder if it'd shift responsibility to whoever runs the agent, if ever relevant later).

And these models can't do basic logical reasoning: https://www.forbes.com/sites/corneliawalther/2025/06/09/intelligence-illusion-what-apples-ai-study-reveals-about-reasoning/  So there might be an effective way to screen contributors. But I understand you may not have the time. Still, I think a rule would be a good idea.

**Comment by linghengqian** [comment]
Incidentally, neither the ASF nor Flatpak mandates that contributors sign a CLA. The ASF only requires project committers and PMC members to sign CLAs. Therefore, to achieve this standard, it would be best if someone pushed through a new issue to require all past and new Flatpak contributors to sign CLAs, thus transferring responsibility.

**Comment by ell1e** [comment]
I'm not sure if a CLA is required. I can't speak on this since I'm not a lawyer, other than I observe projects e.g. use a DCO instead. Or they may just ask you to check a box that you've read the contribution guidelines, or something.

*(Note: I now [brought up gen AI with the Apache Foundation as well](https://github.com/apache/www-site/issues/628).)*

**Comment by ell1e** [comment]
Regarding https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md I can't see any section talking about copyright, or providing any sort of assurances that code you submit is your own, or that code shouldn't be submitted maliciously and should be in a state you believe is operational. Does Flatpak have any such code guidelines? Perhaps that would be a good place to talk about AI policies, if any were to be set into place. Whether that involve a CLA or not.

**Comment by linghengqian** [comment]
> *(Note: I now [brought up gen AI with the Apache Foundation as well](https://github.com/apache/www-site/issues/628).)*


Although this is a bit off-topic... the ASF generally doesn't discuss license issues on GitHub. The ASF uses GitHub only because GitHub is willing to donate infrastructure. Most discussions about licenses take place on https://issues.apache.org/jira/browse/LEGAL and https://the-asf.slack.com/ ; those without an `@apache.org` email address need to register a separate account. By the way, you do have a way to contact the ASF's duty lawyer, but I haven't tried it in any special cases that require legal intervention. See https://www.apache.org/legal/#communications and https://lists.apache.org/list.html?legal-discuss@apache.org .

I don't mind if someone tries to argue with the ASF lawyers in ASF JIRA, after all, I'm not a lawyer.🤣

The prevailing view here is very clear: if AI behaves like a human, any restrictions on AI are meaningless. What used to limit a bad pull request from a human can still limit a bad pull request from an AI, as long as humans or AI are still using Git.

**Comment by ell1e** [comment]
> The prevailing view here is very clear: if AI behaves like a human, any restrictions on AI are meaningless. What used to limit a bad pull request from a human can still limit a bad pull request from an AI, as long as humans or AI are still using Git.

My apologies, I'm not sure what you mean. If I understand you correctly, then it seems to be an orthogonal issue. Bad pull request defenses are useful, but copyright concerns with gen AI don't depend on a pull request being bad.

**Comment by linghengqian** [comment]
What I mean is, unless someone proves in court that the GPL is contagious to AI, then the role of AI in this process is no different from that of an IDE that doesn't use AI. When users apply IntelliJ IDEA's suggested template code changes, they are unlikely to associate IntelliJ IDEA's suggestions with copyright disputes.

If you find any legal ambiguity in https://www.apache.org/legal/generative-tooling.html written by ASF legal professionals, then you should really seek out lawyers, not me. 

You could even consider challenging Oracle in court, given their large legal team and the fact that Oracle has long used GPT-5 to generate pull requests in its open-source projects. See https://github.com/oracle/graalvm-reachability-metadata/pulls?q=is%3Apr+label%3AGenAI+is%3Aclosed .

**Comment by ell1e** [comment]
I'm not a lawyer, but the ASF policy seems potentially practically unfullfilable by actual devs, hence window-dressing. I'm basing this on what I saw smarter people than me say. Basically, things looks complex. Anyway, this isn't legal advice.

**Comment by linghengqian** [comment]
I don't see anything that can be adjusted on the ASF's webpage, because no one has proposed a more feasible solution than the existing one. Unless a particularly ingenious GitHub actions job or web hook suddenly appears, these tools can find which code snippets in PRs are referenced from elsewhere. Currently, the ASF explicitly prohibits copying code from https://stackoverflow.com/questions , which is a similar referencing scenario I'm referring to.

**Comment by ell1e** [comment]
Neither copying StackOverflow nor any other plagiarized code source, like presumably gen AI output, will be reliably caught by most projects. Yet, many are happy to ban both, hence why I thought I'd bring it up. Anyway, I have nothing more to add.

**Comment by ell1e** [comment]
Interesting recent event related to the gen AI licensing chaos: https://www.theregister.com/2026/03/06/ai_kills_software_licensing/

**Comment by linghengqian** [comment]
> Interesting recent event related to the gen AI licensing chaos: https://www.theregister.com/2026/03/06/ai_kills_software_licensing/

In fact, I don't think https://github.com/chardet/chardet/releases/tag/7.0.0 is very helpful for flatpak. The main maintainers of flatpak do not use Agent. Unless I've missed something again.

Furthermore, judging from the course of events, since no lawyer has stepped forward to sue the current maintainers of Chardet, it indicates that Chardet 7.0 is legitimate. Cleanroom development is now considered a standard service; otherwise, someone would have already sued https://malus.sh/ for its `Clean Room as a Service` proposal.

**Comment by ell1e** [comment]
It would surprise me if this was how it worked, nobody sued a project specific (yet) so it supposedly becomes legal for everyone...

**Comment by ell1e** [comment]
After some more thought, `malus.sh` seems like **the** reason for anybody to stop using gen AI altogether. Is this the sort of activity anyone in FOSS wants to be associated with, stripping all projects of any attribution and licensing with a press of a button? Even if it were legal, which I honestly don't know but would find shocking if true, what about respecting the wishes of all other contributors and coders in the field? Gen AI feels like a [disease](https://metr.org/blog/2025-07-10-early-2025-ai-experienced-os-dev-study/) [ruining](https://arstechnica.com/security/2026/01/overrun-with-ai-slop-curl-scraps-bug-bounties-to-ensure-intact-mental-health/) [FOSS](https://www.404media.co/vibe-coding-is-killing-open-source-software-researchers-argue/).

**Update**: seems like `malus.sh` is [satire](https://malus.sh/blog.html) but based on a similar trend done by [others](https://writings.hongminhee.org/2026/03/legal-vs-legitimate/).

**Comment by linghengqian** [comment]
> After some more thought, `malus.sh` seems like the reason for anybody to stop using gen AI altogether. 

Since a significant portion of AI applications are actually being developed for the future, perhaps in six months there will indeed be models capable of supporting such services.

While some communities mentioned in the cited article do clearly oppose AI programming, this doesn't represent the opinion of all foundations, at least not the ASF's. The ASF is still researching and developing new collaboration guidelines and engineering practices to adapt to this new world. After all, `The Apache Way`, promoted by the ASF, has largely helped the ASF maintain a core theoretical framework for the future of Agentic Coding.


I'm not entirely sure about the Linux Foundation's opinion, but since `Agents.md` and `MCP` are now both within the Linux Foundation, I think that represents a certain stance.

The current issue still largely depends on the tastes of the flatpak maintainers.

**Comment by ell1e** [comment]
For what it's worth, I found out recently the plagiarism of LLMs seems to possibly not only be inherent but also seems to get worse with larger models and as they try to get better:
 
1. Practical example of how avoiding hallucinations apparently leads to plagiarism: https://lcamtuf.substack.com/p/large-language-models-and-plagiarism (I linked this one before above.)
 
2. Study looking at model size and larger more accurate models apparently plagiarizing more: https://openreview.net/forum?id=TatRHT_1cK "On the whole, we find that memorization in LMs is more prevalent than previously believed and will likely get worse as models continues to scale, at least without active mitigations."

3. Another study looking at quality being tied to plagiarism: https://www.sciencedirect.com/science/article/pii/S2949719123000213#b7 "In this work we explored the relationship between discourse quality and memorization for LLMs. We found that the models that consistently output the highest-quality text are also the ones that have the highest memorization rate. "

I don't know, it just looks pretty bad.

**Comment by smcv** [comment]
I would suggest that non-maintainers having a detailed discussion of AI policy with other non-maintainers is unlikely to lead to many new insights. This is an issue tracker, not a discussion forum.

You have made your point, and now it's up to the Flatpak team to decide whether it should result in a policy change or not. This is less likely to happen if busy maintainers start unsubscribing from this thread as generating too many notifications.
